cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py

```diff
@@ -0,0 +1,317 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Synthesis for compiling a MPS to a circuit."""
+
+from __future__ import annotations
+
+import logging
+from typing import Literal, TYPE_CHECKING
+
+import numpy as np
+import quimb.tensor as qtn
+
+import cirq
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+
+logger = logging.getLogger(__name__)
+
+
+def gram_schmidt(matrix: NDArray[np.complex128]) -> NDArray[np.complex128]:
+    """Perform Gram-Schmidt orthogonalization on the columns of a matrix
+    to define the unitary block to encode the MPS.
+
+    Notes
+    -----
+    If a column is (approximately) zero, it is replaced with a random vector.
+
+    Args:
+        matrix (NDArray[np.complex128]): Input matrix with complex entries.
+
+    Returns:
+        unitary (NDArray[np.complex128]): A unitary matrix with orthonormal columns
+            derived from the input matrix. If a column is (approximately) zero, it
+            is replaced with a random vector.
+    """
+    num_rows, num_columns = matrix.shape
+    unitary = np.zeros((num_rows, num_columns), dtype=np.complex128)
+    orthonormal_basis: list[NDArray[np.complex128]] = []
+
+    for j in range(num_columns):
+        column = matrix[:, j]
+
+        # If column is (approximately) zero, replace with random
+        if np.allclose(column, 0):
+            column = np.random.uniform(-1, 1, num_rows)  # type: ignore
+            if np.iscomplexobj(matrix):
+                column = column + 1j * np.random.uniform(-1, 1, num_rows)
+
+        # Gram-Schmidt orthogonalization
+        for basis_vector in orthonormal_basis:
+            column -= (basis_vector.conj().T @ column) * basis_vector
+
+        # Handle near-zero vectors (linear dependence)
+        norm = np.linalg.norm(column)
+        if norm < 1e-12:  # pragma: no cover
+            is_complex = np.iscomplexobj(matrix)
+            column = np.random.uniform(-1, 1, num_rows)  # type: ignore
+            if is_complex:
+                column += 1j * np.random.uniform(-1, 1, num_rows)
+            for basis_vector in orthonormal_basis:
+                column -= (basis_vector.conj().T @ column) * basis_vector
+
+        unitary[:, j] = column / np.linalg.norm(column)
+        orthonormal_basis.append(unitary[:, j])
+
+    return unitary
+
+
+class Sequential:
+    def __init__(
+        self, max_fidelity_threshold: float = 0.95, convention: Literal["lsb", "msb"] = "lsb"
```

Comment:
The `msb` convention does not seem to be tested nor handled in the code.

Can we for now assume the `lsb` convention only and drop the `convention` argument?